### ISSUE TITLE
How do you properly pass arguments to JupyterHub spawner json on REST API call?

### ISSUE THREAD

**Issue by wlmvp** [jupyterhub/jupyterhub#2717]
I’m trying to start a single-user server via the REST API without using the default configuration. The [documentation ](https://jupyterhub.readthedocs.io/en/latest/_static/rest-api/index.html#path--users--name--servers--server_name-)1 says this should be possible:

Spawn options can be passed as a JSON body when spawning via the API instead of spawn form. The structure of the options will depend on the Spawner’s configuration.

However, there is no documentation that says how that JSON should be structured. Does such documentation exist anywhere? I want to be able to be able to customize the spawner like I can in using JupyterHub on Kubernetes by changing the config.yaml to have something like:

singleuser:
  image:
    name: jupyter/scipy-notebook
    tag: 59b402ce701d
  cpu:
    guarantee: 0.25
    limit: 0.5
  memory:
    guarantee: "256M"
    limit: "320M"
  profileList:
    - display_name: "Default"
      description: "0.25 CPU; 256M Ram"
      default: True
    - display_name: "BIG"
      description: "0.5 Whole CPUs, 512M Ram"
      kubespawner_override:
        cpu_guarantee: 0.5
        cpu_limit: 0.75
        mem_guarantee: "512M"
        mem_limit: "640M"

tl;dr: How do I structure a JSON for a named-server endpoint POST request to build a server with non-default CPU/memory specs and non-default images? And is there an example in any documentation of doing so that I simply haven’t encountered?

@minrk

**Comment by wlmvp** [comment]
{
      “kubespawner_override”: {
           “image”: “jupyterhub/k8s-singleuser-sample:0.8.2”
     }
}
not effected

**Comment by rkdarst** [comment]
The API doesn't set arbitrary properties on the spawner, it goes through the options form.  Basically, it sets the passed dict to `spawner.user_options`, and then the method `Spawner.load_user_options` loads it. In kubespawner, this can only select a pre-defined profile via `{"profile": display_name}`.  I could be wrong in the exact internal mechanics because it has recently changed, but that's the general way spawners "should" work.

Someone could create a spawner wrapper that does set arbitrary options.

Why is this?  Having a default configuration where any user can override any limits set by the user wouldn't be very secure and not be able to contain the users.

**Comment by wlmvp** [comment]
Thanks for your response, @rkdarst . Does it mean I can't set any extra properties for when i start my server by rest api,  like some extra env variables or set the cpu limit.  And I only select the profile from the profile_list I previously set in config.yaml.

**Comment by rajashreetutgirl** [comment]
@wlmvp Have you got the solution for this? Even I have a similar scenario where I need to pass extra property called "lifecycle_hook" along with request_body of the post request to start the notebook instance.

Thanks!

**Comment by manics** [comment]
I'm closing this as it's quite old and it's a support issue. If you're still having problems please post on the [Jupyter Community Forum](https://discourse.jupyter.org/). Thanks!

**Comment by prateekkhera92** [comment]
@wlmvp @rjshree - looking for the sae thing, any solution you found?
